src/pytorch_lightning/utilities/fetching.py

```diff
@@ -276,14 +276,18 @@ def fetching_function(self) -> Any:
 
     def _fetch_next_batch(self, iterator: Iterator) -> None:
         start_output = self.on_fetch_start()
-        batch = next(iterator)
-        self.fetched += 1
-        if not self.prefetch_batches and self._has_len:
-            # when we don't prefetch but the dataloader is sized, we use the length for `done`
-            dataloader = self.dataloader
-            assert isinstance(dataloader, Sized)  # `_has_len` is True
-            self.done = self.fetched >= len(dataloader)
-        self.on_fetch_end(batch, start_output)
```

Comment:
I would try moving the `on_fetch_start` and `on_fetch_end` into `prefetching`, where we already catch the StopIteration

Reply:
You are correct. But I would only wrap the `batch = next(iterator)` call, not the entire function.

Can you add a test in `tests/tests_pytorch/utilities/test_fetching.py`? It shouldn't use a trainer etc, just creating this class and faking that a stopiteration is raised with some assertions on a mock profiler would work